test/test_invalid_headers.py

```diff
@@ -239,20 +241,33 @@ class TestFilter(object):
     ]
 
     hdr_validation_combos = [
-        h2.utilities.HeaderValidationFlags(is_client, is_trailer)
-        for is_client, is_trailer in [
-            (True, True),
-            (True, False),
-            (False, True),
-            (False, False)
-        ]
+        h2.utilities.HeaderValidationFlags(
+            is_client, is_trailer, is_response_header
+        )
+        for is_client, is_trailer, is_response_header in (
+            itertools.product([True, False], repeat=3)
+        )
     ]
 
     hdr_validation_no_trailers = [
-        h2.utilities.HeaderValidationFlags(is_client, is_trailer)
-        for is_client, is_trailer in [
-            (True, False),
-            (False, False)
+        h2.utilities.HeaderValidationFlags(
+            is_client, is_trailer, is_response_header
+        )
+        for is_client, is_trailer, is_response_header in [
+            (True, False, False),
+            (False, False, False),
+        ]
+    ]
```

Comment:
So the [latest build](https://travis-ci.org/python-hyper/hyper-h2/jobs/160480020#L419) shows the issue a bit clearer I think. @Lukasa I know we want to be checking for the broadest set of options here in `TestFilter` and I think that's why this test may not belong in this specific class. It is hard coded to a set of headers that by definition cannot be a response according to the HTTP/2 spec (It contains request pseudo-headers). 

Since we're already testing that the failure happens in `test_response_header_without_status`, adding a `pytest.raises()` block for every combination that fundamentally cannot work for the given header set is a bit redundant.

I completely agree that we should be testing for all cases, I'm just afraid I may have accidentally started a trend of building unnecessary infrastructure (see: `hdr_validation_no_trailers`) around making a test pass, rather than fixing the underlying problem.